Fix segfault gh 6922

charris opened the pull request:
Revert 23901aa and add regression test for reported segfault.

This is the easy fix for #6922. The style cleanup went just a bit too far and it isn't obvious to me where the problem lies.

charris committed:
TST: Add regression test for gh-6922.

Sorting and unaligned void type should not segfault.

njsmith commented:
LGTM. Do you want to merge this immediately, or is it a strawman to try and nudge @jaimefrio into figuring out the real problem? :-)

charris commented:
I'm going for the quick fix and backport, @jaimefrio is stacked behind google doctrine at the moment. Will probably release a 1.10.3 in a few days.

njsmith commented:
Well, +1 pending CI

njsmith commented:
Actually, looks like you've got test failures on py3

charris commented:
Grrr. Should be fixed, just need the array first field filled with something better than None objects.

njsmith commented:
LGTM

seberg commented:
The dummy trick part in the new code looks very nice. The other part of the code is incorrect, because the OBJECT_copyswap function has to xdecref the destination. And the cache is not NULL'ed. The best fix would be to only undo that part (and add a comment of why maybe).

seberg commented:
i am fine with reverting though, was just to note. The change looks nice, but copyswap just does not work here for object type. Maybe @jaimefrio has an idea later, can't think of anything that is better than the old code right now.

njsmith commented:
@seberg: Maybe the easiest way forward would be to merge this and then redo
the parts that make sense in a new commit?

On Sat, Jan 2, 2016 at 2:11 PM, seberg <notifications@github.com> wrote:

> i am fine with reverting though, was just to note. The change looks nice,
> but copyswap just does not work here for object type. Maybe @jaimefrio
> <https://github.com/jaimefrio> has an idea later, can't think of anything
> that is better than the old code right now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6924#issuecomment-168434671>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

seberg commented:
I don't care, you or Chuck can do whichever you like. The sufficient change I think is just this (twice, a second one is basically identical):
```
-                new->f->copyswap(nip2, ip2 + offset, swap, dummy);
 +                memcpy(nip2, ip2 + offset, new->elsize);
 +                if (swap)
 +                    new->f->copyswap(nip2, NULL, swap, dummy);
```

charris commented:
@seberg Yes, I liked the cleanup, just didn't want to spend a lot of time finding which parts to fix. I'm going to put this, but would welcome a better fix on top of this.

charris commented:
@seberg That seems to work. I'll redo this PR.

charris committed:
BUG: #6922:  Fix segfault introduced in 23901aa.

Revert troublesome parts of gh-5929. Copyswap cannot be relied
upon for void types containing objects.

seberg commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2866,7 +2866,9 @@ VOID_compare(char *ip1, char *ip2, PyArrayObject *ap)
                     goto finish;
                 }
-                new->f->copyswap(nip1, ip1 + offset, swap, dummy);
+                memcpy(nip2, ip2 + offset, new->elsize);
```
oops, it is ip1 here and nip1.

seberg commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2866,7 +2866,9 @@ VOID_compare(char *ip1, char *ip2, PyArrayObject *ap)
-                new->f->copyswap(nip1, ip1 + offset, swap, dummy);
+                memcpy(nip2, ip2 + offset, new->elsize);
+                if (swap)
+                    new->f->copyswap(nip2, NULL, swap, dummy);
```
Same here then of course.

charris commented:
Clearly we need better tests ;) And less rush.

seberg commented:
Hehe, the tests found it, otherwise I would not have noticed! (though I doubt they would find the swap part).

charris commented:
Hmm, didn't show up on my home machine.

seberg commented:
Well, that is what we have Travis for :). Anyway, thanks! Going to put this in then. (AppVeyor not done, but since we ignore it anyway)

jaimefrio commented:
Wow, thanks for the quick and thorough clean-up of my mess up!

After looking into it, I think it would have preferred explicitly zeroing the cache when needed, i.e. something along the lines of:

```C
if (PyDataType_REFCHK(new)) {
    /* The cache will be XDECREF'ed by copyswapn, must init it to NULLs */
    nip1 = npy_alloc_cache_zero(new->elsize);
}
else {
    nip1 = npy_alloc_cache(new->elsize);
}
if (nip1 == NULL) {
    goto finish;
}
new->f->copyswap(nip1, ip1 + offset, swap, dummy);
```

But I guess `memcpy` is also a good enough option.
